MKL: Level-1 (scal) primitive support

kballeda opened the pull request:
This PR consists of Level-1 (scal) primitive support for oneJulia

kballeda committed:
cuBlas level 1 method - scal supported

kballeda committed:
scal test case updated in onemkl.jl

kballeda committed:
NITS - cleanup

kballeda committed:
indentation NITS

kballeda committed:
updated with scal - deps

kballeda committed:
indentation fixes

kballeda committed:
NITS

kballeda committed:
NITS

kballeda committed:
cleanup onemkl.jl

pengtu reviewed (commented):
There is a 'testf' function defined in test/setup.jl, which uses the TestSuite.compare function in
https://github.com/JuliaGPU/GPUArrays.jl/blob/master/test/testsuite.jl#L38.  It copies the input data to CPU and GPU arrays and compare the results.  Please modify the test using the 'testf' function for more concise code.

kballeda committed:
updated with rmul! and testf usage

kballeda committed:
NITS

kballeda commented:
> There is a 'testf' function defined in test/setup.jl, which uses the TestSuite.compare function in https://github.com/JuliaGPU/GPUArrays.jl/blob/master/test/testsuite.jl#L38. It copies the input data to CPU and GPU arrays and compare the results. Please modify the test using the 'testf' function for more concise code.

Thanks for your inputs. I have implemented rmul! and testf based tests they fail for one case commonly. 
If I check the oneMKL.scal primitive for CPU/GPU tests are passing. I have added the changes please take look when you get a chance.

kballeda committed:
testf used for cpu/gpu testing.
1. Input array is converted to non-gpu array
2. alpha is rand instead of based on T typewq

kballeda committed:
NITS - clenaup & included int specific calls to rmul! diverted to
non-gpuarray flow

kballeda committed:
wrapper alpha turns elttype and support all combinationswq

kballeda committed:
NITS

kballeda committed:
support for Cs, Zd configs of scal function

kballeda committed:
updated with staticcast complex alpha

pengtu commented on lib/mkl/wrappers.jl:
```diff
@@ -14,7 +14,24 @@ function Base.convert(::Type{onemklTranspose}, trans::Char)
+    @eval begin
+    	function scal!(n::Integer,
+                       alpha::Number,
+                       x::StridedArray{$elty})
```
Please change this to oneStridedArray{$elty} as Tim suggested in a previous PR.

kballeda replied:
Done, Thanks

pengtu commented on lib/mkl/linalg.jl:
```diff
@@ -49,6 +49,13 @@ function gemm_dispatch!(C::oneStridedVecOrMat, A, B, alpha::Number=true, beta::N
     end
 end
 
+LinearAlgebra.rmul!(x::oneStridedVecOrMat{<:onemklFloat}, k::Number) = 
```
I am not sure if these two rules will properly dispatch the combination of alpha types of ComplexF32, ComplexF64, F32 and F64 to the ComplexF32 and Complex64 scal functions.

Please write specific tests for all the combinations to make sure they are dispatched properly.

If not, we will want to use AMDGPU.jl's type based dispatching rules at
 https://github.com/JuliaGPU/AMDGPU.jl/blob/master/src/blas/wrappers.jl#L85, and

https://github.com/JuliaGPU/AMDGPU.jl/blob/master/src/blas/wrappers.jl#L106

kballeda replied:
Sure, Thanks

pengtu commented on test/onemkl.jl:
```diff
@@ -0,0 +1,16 @@
+@testset "level 1" begin
+	@testset for T in eltypes
+		if T <:oneMKL.onemklFloat
+			@test testf(rmul!, rand(T,m), rand())
```
See the comment above on the dispatching rules.  The rand() function on the 'alpha' only creates 'float' type numbers so not all combinations of 'alpha' - ComplexF32, ComplexF64 are tested in this test.

kballeda replied:
Yes agree, I have reenabled the original rand(T,1) to launch float, complex type tests. The cross combination ComplexF32 [x] vs Float32 [Alpha] scenarios I will write separate tests.

pengtu reviewed (commented):
Please review my comments.  Thanks!

kballeda committed:
added onestridedarray

kballeda committed:
enable tests of complex tye

kballeda committed:
Merge branch 'master' into l1_scal

kballeda committed:
updated with Csscal and Zdscal test enabled

kballeda committed:
NITS

kballeda committed:
NITS

kballeda committed:
Merge branch 'master' into l1_scal

kballeda committed:
NITS

kballeda committed:
NITS

kballeda committed:
Cleanup of tests

kballeda committed:
Instead of isapprox use compare op

kballeda committed:
Bug fix: disable f16 check as it is not supported (CI crash)

kballeda committed:
Merge branch 'master' into l1_scal

kballeda committed:
Merge branch 'master' into l1_scal

kballeda committed:
use force flush instead of wait()
